openapi-codegen/core/src/main/scala/sttp/tapir/codegen/JsonSerdeGenerator.scala

```diff
@@ -0,0 +1,286 @@
+    val uncapitalisedName = name.head.toLower +: name.tail
+    s"""
+       |implicit lazy val ${uncapitalisedName}JsonCodec: $jsoniterPkgCore.JsonValueCodec[${name}] = $jsoniterPkgMacros.JsonCodecMaker.make($jsoniteEnumConfig.withDiscriminatorFieldName(scala.None))""".stripMargin
+  }
+
+  private def genJsoniterNamedSerde(name: String): String = {
+    val uncapitalisedName = name.head.toLower +: name.tail
+    s"""
+       |implicit lazy val ${uncapitalisedName}JsonCodec: $jsoniterPkgCore.JsonValueCodec[$name] = $jsoniterPkgMacros.JsonCodecMaker.make($jsoniterBaseConfig)""".stripMargin
+  }
+
+  private def generateJsoniterAdtSerde(
+      name: String,
+      schema: OpenapiSchemaOneOf,
+      maybeFullModelPath: Option[String]
+  ): String = {
+    val fullPathPrefix = maybeFullModelPath.map(_ + ".").getOrElse("")
+    val uncapitalisedName = name.head.toLower +: name.tail
+    schema match {
+      case OpenapiSchemaOneOf(_, Some(discriminator)) =>
+        val schemaToJsonMapping = discriminator.mapping
+          .map { case (jsonValue, fullRef) => fullRef.stripPrefix("#/components/schemas/") -> jsonValue }
+        val body = if (schemaToJsonMapping.exists { case (className, jsonValue) => className != jsonValue }) {
+          val discriminatorMap = indent(2)(
+            schemaToJsonMapping
+              .map { case (k, v) => s"""case "$fullPathPrefix$k" => "$v"""" }
+              .mkString("\n", "\n", "\n")
+          )
+          val config =
+            s"""$jsoniterBaseConfig.withRequireDiscriminatorFirst(false).withDiscriminatorFieldName(Some("${discriminator.propertyName}")).withAdtLeafClassNameMapper{$discriminatorMap}"""
+          val serde =
+            s"implicit lazy val ${uncapitalisedName}Codec: $jsoniterPkgCore.JsonValueCodec[$name] = $jsoniterPkgMacros.JsonCodecMaker.make($config)"
+
+          s"""$serde
+             |""".stripMargin
+        } else {
+          val config =
+            s"""$jsoniterBaseConfig.withRequireDiscriminatorFirst(false).withDiscriminatorFieldName(Some("${discriminator.propertyName}"))"""
+          s"implicit lazy val ${uncapitalisedName}Codec: $jsoniterPkgCore.JsonValueCodec[$name] = $jsoniterPkgMacros.JsonCodecMaker.make($config)"
+        }
+        body
+
+      case OpenapiSchemaOneOf(schemas, None) =>
+        val childNameAndSerde = schemas.collect { case ref: OpenapiSchemaRef =>
+          val name = ref.stripped
+          name -> s"${name.head.toLower +: name.tail}JsonCodec"
+        }
+        val childSerdes = childNameAndSerde.map(_._2)
+        val doDecode = childSerdes.mkString("List(\n  ", ",\n  ", ")\n") +
+          indent(2)(s""".foldLeft(Option.empty[$name]) {
+          |  case (Some(v), _) => Some(v)
+          |  case (None, next) =>
+          |    scala.util.Try(next.asInstanceOf[$jsoniterPkgCore.JsonValueCodec[$name]].decodeValue(in, default))
+          |      .fold(_ => { in.rollbackToMark(); None }, Some(_))
```

Comment:
This feels clunky and gross and inefficient, but then I suppose that's why it's a good idea to use discriminators.